tsdb/engine/tsm1/engine.go

```diff
@@ -151,47 +154,81 @@ func (e *Engine) SetEnabled(enabled bool) {
 // all running compactions are aborted and new compactions stop running.
 func (e *Engine) SetCompactionsEnabled(enabled bool) {
 	if enabled {
-		e.mu.Lock()
-		if e.compactionsEnabled {
-			e.mu.Unlock()
-			return
-		}
-		e.compactionsEnabled = true
+		e.enableSnapshotCompactions()
+		e.enableLevelCompactions()
 
-		e.done = make(chan struct{})
-		e.Compactor.Open()
+	} else {
+		e.disableSnapshotCompactions()
+		e.disableLevelCompactions()
+	}
+}
 
+func (e *Engine) enableLevelCompactions() {
+	e.mu.Lock()
+	if e.levelCompactionsEnabled {
 		e.mu.Unlock()
+		return
+	}
+	e.levelCompactionsEnabled = true
+	e.Compactor.EnableCompactions()
+	e.done = make(chan struct{})
+	e.mu.Unlock()
 
-		e.wg.Add(5)
-		go e.compactCache()
-		go e.compactTSMFull()
-		go e.compactTSMLevel(true, 1)
-		go e.compactTSMLevel(true, 2)
-		go e.compactTSMLevel(false, 3)
-	} else {
-		e.mu.Lock()
-		if !e.compactionsEnabled {
-			e.mu.Unlock()
-			return
-		}
-		// Prevent new compactions from starting
-		e.compactionsEnabled = false
+	e.wg.Add(4)
+	go e.compactTSMFull()
+	go e.compactTSMLevel(true, 1)
+	go e.compactTSMLevel(true, 2)
+	go e.compactTSMLevel(false, 3)
+}
+
+func (e *Engine) disableLevelCompactions() {
+	e.mu.Lock()
+	if !e.levelCompactionsEnabled {
 		e.mu.Unlock()
+		return
+	}
+	// Prevent new compactions from starting
+	e.levelCompactionsEnabled = false
+	e.Compactor.DisableCompactions()
+	e.mu.Unlock()
+
+	// Stop all background compaction goroutines
+	close(e.done)
+
+	// Wait for compaction goroutines to exit
+	e.wg.Wait()
+
+	if err := e.cleanup(); err != nil {
+		e.logger.Printf("error cleaning up temp file: %v", err)
+	}
+}
 
-		// Stop all background compaction goroutines
-		close(e.done)
+func (e *Engine) enableSnapshotCompactions() {
+	e.mu.Lock()
+	if e.snapshotCompactionsEnabled {
+		e.mu.Unlock()
+		return
+	}
 
-		// Abort any running goroutines (this could take a while)
-		e.Compactor.Close()
+	e.snapshotCompactionsEnabled = true
+	e.snapshotterDone = make(chan struct{})
```

Comment:
`snapshotterDone` as far as I can see is never sent on or closed, so I'm not sure what its purpose it?